### DISCUSSION TITLE
how to write/ add a callback modules?

### DISCUSSION THREAD

**Discussion by synapticonashaikh** [micropython/micropython#10188] [Core Development]
Hello all, 
I'm working on a project where I'm porting the Micropython stack on one of the SOC boards. As the support library of Micropython for the SOC (XMOS based) I'm using is not listed, I used the ports/minimal project. Now, module by module, I'm developing the modules that can provide access to the hardware layer. So, now moving further, I want to build a callback module which takes a user function and triggers based on the GPIO interrupt. In short, just like any ARM-based GPIO interrupt. 

Therefore, I would like to request you all help me in creating a callback function based on interrupt. 

Micopython sample code: 

```
from machine import Pin
def callback(p):
    print('pin change', p)
p0 = Pin(0, Pin.IN)
p0.irq(trigger=Pin.IRQ_FALLING, handler=callback)
while True: 
    pass
```

**Comment by jimmo** [comment]
@synapticonashaikh You should look at how the other ports implement exactly this (e.g. for pin IRQs). ports/samd/machine_pin.c is a good example (but just about all ports do this). Let us know if you have any specific questions about how this is done.

**Reply by synapticonashaikh** [reply]
I'm trying to go through these files, but it is a bit complicated for me to understand. Also, the interrupt in XMOS works differently than the regular interrupt. Unlike the ARM and other controllers, XMOS doesn't have a handler function for interrupts. Instead, there is a macro `DEFINE_INTERRUPT_CALLBACK`, which defines the callback space, and `DEFINE_INTERRUPT_PERMITTED` defines the permitted area where interrupts are allowed (You can see in the code sample). 

Therefore, what I wanted to do whenever I call the `irq(handler=callback)` code should store the handler and execute that handler inside the `DEFINE_INTERRUPT_CALLBACK` block (whenever interrupts occur).

```
/* Wrapper C file */
/* Includes */
  #include <stdio.h>
  #include <xcore/hwtimer.h>
  #include <xcore/triggerable.h>
  #include <xcore/port.h>
  #include <xcore/interrupt.h>
  #include <xcore/interrupt_wrappers.h>

/* External Functions */ 
  extern void CallbackFunction(void);
  extern void FnParallel(void);

/* GLOBAL VARIABLE DECLARATION */
  port_t button1 = XS1_PORT_1P;
  static uint8_t RisingFallingEdge; 
  static uint8_t uifeedback,uiStatus;

/* Interrupt Definitions*/
DEFINE_INTERRUPT_PERMITTED(interrupt_handlers, void, interruptable_task, void)
{
  interrupt_unmask_all( ); 
  FnParallel( );
}

DEFINE_INTERRUPT_CALLBACK (interrupt_handlers, interrupt_task, button)
{
  //manipulation to get the interrupt on rising edge only
  port_set_trigger_in_not_equal(button1, RisingFallingEdge);
  RisingFallingEdge = !RisingFallingEdge;
  uifeedback = port_peek(button1); 
  if      (( uifeedback == 1 ) && (uiStatus == 0))
          { CallbackFunction( ); uiStatus = 1;}
  else if ((uifeedback == 0 ) && (uiStatus == 1)) 
          { uiStatus = 0;   }
}

/* Function definition */
void GPIOINTRWrapper(void)
{ INTERRUPT_PERMITTED(interruptable_task)( ); }

void GPIOInterrupt(void)
{
  port_enable(button1);
  triggerable_setup_interrupt_callback(button1, &button1, INTERRUPT_CALLBACK(interrupt_task));
  port_set_trigger_in_not_equal(button1, 1);
  port_clear_trigger_in(button1);
  triggerable_enable_trigger(button1);
}

```
```
/* hedaers */
/* main file*/
  #include <platform.h>
  #include <xs1.h>
  #include <timer.h>
  #include <flash.h>
  #include <print.h>    

/* External Function */  
  extern "C" {  void GPIOInterrupt(void); void GPIOINTRWrapper(void); }        

/* GLOBAL VARIABLE DECLARATION */
    timer stTime1,stTime2;    
    int CallbackCount = 0;

/* Fnction Definitions */
void CallbackFunction(void)
{
    CallbackCount++;
    printf ("CALLBACK FUNCTION=%d\n\r",CallbackCount);
}
void FnParallel(void)
{
  uint64_t uiTimeTotal; uint32_t uiCount=0; 
  stTime1 :> uiTimeTotal; uiTimeTotal = uiTimeTotal + ui1Sec ;   
  while (SET)
  {
      stTime1 when timerafter(uiTimeTotal) :> void;    
      uiTimeTotal = uiTimeTotal + ui1Sec ; uiCount++;
      printf("1S=%u\n\r",uiCount);    
  }  
}
int main ( )
{
  GPIOInterrupt( );
  GPIOINTRWrapper( );
 /*control should not reach here*/   
 return RESET;   
}

```

**Reply by jimmo** [reply]
> Therefore, what I wanted to do whenever I call the irq(handler=callback) code should store the handler and execute that handler inside the DEFINE_INTERRUPT_CALLBACK block (whenever interrupts occur).

Rather than trying to make the Python function the actual interrupt handler, you need to always have a C function that is the registered handler, and then you can call the Python function from there (or probably safer, use mp_sched_schedule to schedule the python callback).

**Reply by synapticonashaikh** [reply]
Hello, I tried to understand the functionality in the `ports/samd/machine_pin.c`, and I copied and pasted the irq function `STATIC mp_obj_t gpiopin_irq(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args)` in one of the existing modules. But while running, I keep getting this error `NameError: name not defined`, and I don't know how to solve this.

----->IRQ function
```
//pin.irq(handler=None, trigger=IRQ_FALLING|IRQ_RISING, hard=False)
STATIC mp_obj_t gpiopin_irq(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) 
{
    printf("in the gpiopin_irq function \n\r");
    enum { ARG_handler, ARG_trigger, ARG_hard };
    static const mp_arg_t allowed_args[] = {
        { MP_QSTR_handler, MP_ARG_OBJ, {.u_rom_obj = MP_ROM_NONE} },
        { MP_QSTR_trigger, MP_ARG_INT, {.u_int = 3} },
        { MP_QSTR_hard, MP_ARG_BOOL, {.u_bool = false} },
    };
    pin_obj_t *self = MP_OBJ_TO_PTR(pos_args[0]);
    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);

    if (n_args > 1 || kw_args->used != 0) 
    {    }

    // TODO should return an IRQ object
    return mp_const_none;
}
STATIC MP_DEFINE_CONST_FUN_OBJ_KW(pin_irq_obj, 1, gpiopin_irq);
```
----->Added QSTR definition
```
  { MP_ROM_QSTR(MP_QSTR_pirq),       MP_ROM_PTR(&pin_irq_obj) },
  { MP_ROM_QSTR(MP_QSTR_IRQ_RISING),  MP_ROM_INT(IRQ_FALLING) },
  { MP_ROM_QSTR(MP_QSTR_IRQ_FALLING), MP_ROM_INT(IRQ_RISING) },
```
----->Script
```
import gpio
def Func(self):
    pass
val=gpio.pirq(handler=Func, trigger=IRQ_FALLING|IRQ_RISING, hard=False)
```
----->Error
```
Traceback (most recent call last):
  File "<stdin>", in <module>
NameError: name not defined
```

**Reply by jimmo** [reply]
@synapticonashaikh 

In `val=gpio.pirq(handler=Func, trigger=IRQ_FALLING|IRQ_RISING, hard=False)`

you are using the `IRQ_FALLING` and `IRQ_RISING` constants directly. You need to write `gpio.IRQ_FALLING`

**Reply by synapticonashaikh** [reply]
After replacing the `IRQ_FALLING` with `gpio.IRQ_FALLING`, I'm getting the following error. Rest everything is same
```
Traceback (most recent call last):
  File "<stdin>", in <module>
TypeError: argument num/types mismatch
```

**Reply by jimmo** [reply]
You have copied the argument handling code from a function that has a `self` argument, and so it's offsetting the arguments by one:

`mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);`

Your method is just directly on the module, not on a class, so it has no self argument. Remove the two `- 1` and it should work.

**Reply by synapticonashaikh** [reply]
I practically tried all the combinations, but the result was the same. 
I removed `self` from the script, and in `n_args - 1, pos_args + 1` removed `-1, +1`. Likewise did all the combinations.
I notice that `printf("in the gpiopin_irq function \n\r");` is not executing, which might mean that due to syntactical error, it is not entering the function at all.

**Reply by jimmo** [reply]
You have `STATIC MP_DEFINE_CONST_FUN_OBJ_KW(pin_irq_obj, 1, gpiopin_irq);` whcih says this function must take one positional argument. But you're calling it with entirely keyword arguments. Change the 1 to a 0.
